File: lab3/python/dh_setup.py
from galois_field import GaloisField

from copy import deepcopy
from random import randint

from typing import Generic, TypeVar, List

T = TypeVar("T", bound=GaloisField)

class DHSetup(Generic[T]):
    def __init__(self):
        self.generator = GaloisField(0) # set temporarly to be able to call characteristic()
        primes = self._getPrimeDivisors()
        g = self._getRandom()
        while not self._isGenerator(g, primes):
            g = self._getRandom()
        self.generator = g

    def getGenerator(self) -> T:
        return self.generator
    
    def power(self, base: T, exponent: int) -> T:
        if exponent == 0:
            return GaloisField(1)
        
        baseCopy = deepcopy(base)
        result = GaloisField(1)
        

        while exponent > 0:
            if exponent % 2 == 1:
                result = result * baseCopy
                exponent = exponent - 1
            baseCopy = baseCopy * baseCopy
            exponent = exponent // 2

        return result
    
    def _getRandom(self) -> T:
        return GaloisField(randint(2, self.generator.characteristic() - 1))
    
    def _getPrimeDivisors(self) -> List[int]:
        divisors = []
        n = self.generator.characteristic() - 1
        i = 2
        while i * i <= n:
            if n % i == 0:
                divisors.append(i)
                while n % i == 0:
                    n //= i
            i += 1
        return divisors
    
    def _isGenerator(self, g: T, primes: List[int]) -> bool:
        for p in primes:
            if self.power(g, (self.generator.characteristic() - 1) // p) == GaloisField(1):
                return False
        return True

File: lab3/java/src/main/java/com/diffiehellman/DHSetup.java
package com.diffiehellman;

import java.util.ArrayList;
import java.util.Random;

import com.diffiehellman.GaloisField;

public class DHSetup<T extends GaloisField> {
    private static Random random = new Random();
    private T generator;

    public DHSetup() {
        this.generator = (T) new GaloisField(2);
        long characteristic = this.generator.getCharacteristic();
        final ArrayList<Long> primeDivisors = this.generatePrimeDivisors(characteristic - 1);
        Random random = new Random();
        do {
            this.generator = (T) new GaloisField(Math.abs(random.nextLong(characteristic - 2) + 1));
        } while(!this.isGenerator(this.generator, primeDivisors));
    }

    public T getGenerator() {
        return this.generator;
    }

    public T power(T base, long exponent) {
        if(exponent == 0) {
            return (T) new GaloisField(1);
        }

        T baseCopy = (T) new GaloisField(base);
        T result = (T) new GaloisField(1);

        while(exponent > 0) {
            if(exponent % 2 == 1) {
                result.multiply(baseCopy);
                exponent--;
            }

            baseCopy.multiply(baseCopy);
            exponent /= 2;
        }

        return result;
    }

    private ArrayList<Long> generatePrimeDivisors(long upperBound) {
        ArrayList<Long> divisors = new ArrayList<>();
        for(long i = 2; i * i <= upperBound; i++) {
            if(upperBound % i == 0) {
                divisors.add(i);
                while(upperBound % i == 0) {
                    upperBound /= i;
                }
            }
        }

        return divisors;
    }

    private boolean isGenerator(final T value, final ArrayList<Long> primeDivisors) {
        for(long divisor : primeDivisors) {
            T power = this.power(value, (value.getCharacteristic() - 1) / divisor);
            if(power.toLong() == 1) {
                return false;
            }
        }

        return true;
    }
}